cmd/backport-verifier/main.go

```diff
@@ -0,0 +1,184 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/sirupsen/logrus"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	prowconfig "k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/config/secret"
+	prowflagutil "k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/githubeventserver"
+	"k8s.io/test-infra/prow/interrupts"
+	"k8s.io/test-infra/prow/logrusutil"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+// Config maps upstreams to downstreams for verification
+type Config struct {
+	// Repositories is a mapping of downstream org/repo to upstream org/repo
+	Repositories map[string]string `json:"repositories,omitempty"`
+}
+
+func (c *Config) validate() error {
+	var errs []error
+	for downstreamRepo, upstreamRepo := range c.Repositories {
+		if len(strings.Split(downstreamRepo, "/")) != 2 {
+			return fmt.Errorf("%s should be in org/repo format", downstreamRepo)
+		}
+
+		if len(strings.Split(upstreamRepo, "/")) != 2 {
+			return fmt.Errorf("%s should be in org/repo format", upstreamRepo)
+		}
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+type options struct {
+	mut *sync.RWMutex
+
+	configPath        string
+	webhookSecretFile string
+
+	config *Config
+
+	githubEventServerOptions githubeventserver.Options
+	github                   prowflagutil.GitHubOptions
+
+	dryRun bool
+}
+
+func gatherOptions() options {
+	o := options{
+		mut: &sync.RWMutex{},
+	}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+
+	fs.StringVar(&o.configPath, "config-path", "", "Path to backport verifier configuration.")
+	fs.StringVar(&o.webhookSecretFile, "hmac-secret-file", "", "Path to the file containing the GitHub HMAC secret.")
+
+	o.github.AddFlags(fs)
+	o.githubEventServerOptions.Bind(fs)
+
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatalf("cannot parse args: '%s'", os.Args[1:])
+	}
+	return o
+}
+
+func (o *options) Validate() error {
+	if err := o.github.Validate(o.dryRun); err != nil {
+		return err
+	}
+
+	bytes, err := ioutil.ReadFile(o.configPath)
+	if err != nil {
+		return fmt.Errorf("couldn't read configuration file: %v", o.configPath)
+	}
+
+	if err := yaml.Unmarshal(bytes, o.config); err != nil {
+		return fmt.Errorf("couldn't unmarshal configuration: %w", err)
+	}
+
+	if err := o.config.validate(); err != nil {
+		return err
+	}
+
+	if err := o.githubEventServerOptions.DefaultAndValidate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *options) getConfigWatchAndUpdate() (func(ctx context.Context), error) {
+	errFunc := func(err error, msg string) {
+		logrus.WithError(err).Error(msg)
+	}
+
+	eventFunc := func() error {
+		bytes, err := ioutil.ReadFile(o.configPath)
+		if err != nil {
+			return fmt.Errorf("couldn't read configuration file %s: %w", o.configPath, err)
+		}
+
+		var c *Config
+		if err := yaml.Unmarshal(bytes, c); err != nil {
```

Comment:
/hold

This definitely doesn't work and if we have code elsewhere that does the same then its similarly broken. It makes total sense as the `yaml.Unmrshal` can not manipulate the pointer address we store in line 116, as it gets a copy of that address.

```
$ cat main.go 
package main

import (
	"fmt"
	"github.com/ghodss/yaml"
	"log"
)

type Config struct {
	// Repositories is a mapping of downstream org/repo to upstream org/repo
	Repositories map[string]string `json:"repositories,omitempty"`
}

func main() {
	var c *Config
	fmt.Printf("%p\n", c)
	raw := []byte(`
repositories:
  hello: world`,
	)
	if err := yaml.Unmarshal(raw, c); err != nil {
		log.Fatalf("unmarshal failed: %v", err)
	}
	re, err := yaml.Marshal(c)
	if err != nil {
		log.Fatalf("Remarshal failed: %v", err)
	}
	println(string(re))
}
```

gives
```
0x0
2021/01/28 16:37:41 unmarshal failed: error unmarshaling JSON: json: Unmarshal(nil *main.Config)
exit status 1
```